controller_manager/test/test_controllers_chaining_with_controller_manager.cpp

```diff
@@ -610,6 +679,271 @@ TEST_P(
+  // the other preceding controller should be active (diff_drive_controller_two)
+  ASSERT_EQ(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE, diff_drive_controller->get_state().id());
+  ASSERT_EQ(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, diff_drive_controller_two->get_state().id());
+}
+
+TEST_P(
+  TestControllerChainingWithControllerManager, test_chained_controllers_deactivation_error_handling)
+{
+  SetupControllers();
+
+  // add all controllers - CONTROLLERS HAVE TO ADDED IN EXECUTION ORDER
+  cm_->add_controller(
+    position_tracking_controller, POSITION_TRACKING_CONTROLLER,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    diff_drive_controller, DIFF_DRIVE_CONTROLLER,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    diff_drive_controller_two, DIFF_DRIVE_CONTROLLER_TWO,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    pid_left_wheel_controller, PID_LEFT_WHEEL,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    pid_right_wheel_controller, PID_RIGHT_WHEEL,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+
+  CheckIfControllersAreAddedCorrectly();
+
+  ConfigureAndCheckControllers();
+
+  // Set ControllerManager into Debug-Mode output to have detailed output on updating controllers
+  cm_->get_logger().set_level(rclcpp::Logger::Level::Debug);
+  rclcpp::get_logger("ControllerManager::utils").set_level(rclcpp::Logger::Level::Debug);
+
+  // at beginning controllers are not in chained mode
+  EXPECT_FALSE(pid_left_wheel_controller->is_in_chained_mode());
+  EXPECT_FALSE(pid_right_wheel_controller->is_in_chained_mode());
+  ASSERT_FALSE(diff_drive_controller->is_in_chained_mode());
+
+  // Activate following controllers
+  ActivateAndCheckController(
+    pid_left_wheel_controller, PID_LEFT_WHEEL, PID_LEFT_WHEEL_CLAIMED_INTERFACES, 1u);
+  ActivateAndCheckController(
+    pid_right_wheel_controller, PID_RIGHT_WHEEL, PID_RIGHT_WHEEL_CLAIMED_INTERFACES, 1u);
+
+  // Test Case 5: Deactivating a preceding controller that is not active --> return error; all controller stay in the same state
+  // There is different error and timeout behavior depending on strictness
+  auto const getExpectedBehaviorVals = [&]
+  {
+    if (
+      test_param.strictness ==
+      controller_manager_msgs::srv::SwitchController::Request::STRICT)  // STRICT
+    {
+      return std::make_tuple(controller_interface::return_type::ERROR, std::future_status::ready);
+    }
+    else if (
+      test_param.strictness ==
+      controller_manager_msgs::srv::SwitchController::Request::BEST_EFFORT)  // BEST EFFORT
+    {
+      return std::make_tuple(controller_interface::return_type::OK, std::future_status::ready);
+    }
+    else
+    {
+      std::runtime_error("test_param was not defined!");
+    }
+  };
+  auto const expected = getExpectedBehaviorVals();
```

Comment:
Also put into a map with structures. It will be easier to handle.

Reply:
Done